### PR TITLE
New Lexer, Parser and Interpreter

### PR DESCRIPTION
**Pull request by joerivanarkel**

### CODE REVIEW

File: NewInterpreterTest/Values/Number.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+public class Number : IValue, INumber
+{
+    public int Value { get; set; }
+
+    public Number(int value)
+    {
+        Value = value;
+    }
+
```
Comment (teusvanarkel) [inline NewInterpreterTest/Values/Number.cs:14]: Maybe move the method to another class

File: NewLexerTest/Error.cs [hunk 2]
```diff
@@ -1,27 +1,20 @@
...
         public string details { get; set; }
 
-        public Error(Position PositionStart, Position PositionEnd, string ErrorName, string Details)
+        public Error(string ErrorName, string Details)
         {
-            positionStart = PositionStart;
-            positionEnd = PositionEnd;
             errorName = ErrorName;
             details = Details;
         }
```
Comment (teusvanarkel) [inline NewLexerTest/Error.cs:14]: Override void ToString()

File: NewLexerTest/Lexer.cs [hunk 3]
```diff
@@ -13,28 +13,29 @@ public class Lexer
     public Lexer(string FileName, string Text)
     {
         fileName = FileName;
-        position = 0;
+        position = -1;
         text = Text.Replace("\n", "").Replace("\r", "").Replace("    ", "");
+        Advance();
     }
 
     public static void Advance()
```
Comment (teusvanarkel) [inline NewLexerTest/Lexer.cs:21]: Remove the statics!!!!!

File: NewLexerTest/TokenRepository.cs [hunk 4]
```diff
@@ -1,14 +1,16 @@
...
+using NewLexerTest.Enums;
+using NewPirateLexer.Enums;
+using NewPirateLexer.Tokens;
 
-namespace PirateLexer
+namespace NewPirateLexer
 {
     public class TokenRepository
     {
         public static Token MakeNumber()
```
Comment (teusvanarkel) [inline NewLexerTest/TokenRepository.cs:10]: remove the statics!!!

File: NewParserTest/Node/BinaryOperationNode.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    public INode Right { get; set; }
+
+    public BinaryOperationNode(INode left, Token operatorToken, INode right)
+    {
+        Left = left;
+        Operator = operatorToken;
+        Right = right;
+    }
+
+    public string Display()
```
Comment (teusvanarkel) [inline NewParserTest/Node/BinaryOperationNode.cs:20]: override void ToString

File: NewParserTest/Node/ComparisonOperationNode.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+using NewPirateLexer.Tokens;
+using NewParserTest.Node.Interfaces;
+
+namespace NewParserTest.Node;
+
+public class ComparisonOperationNode : INode, IOperationNode
+{
+    public INode Left { get; set; }
```
Comment (teusvanarkel) [inline NewParserTest/Node/ComparisonOperationNode.cs:8]: the properties can be privates?

File: NewParserTest/Parser.cs [hunk 7]
```diff
@@ -1,13 +1,38 @@
...
+        Scope scope = new Scope();
+        while (index + 1 != _tokens.Count())
+        {
+            if(_tokens == null) throw new ArgumentNullException(nameof(_tokens));
+            var tokenParser = parserFactory.GetParser(_tokens[index], _tokens);
+            var parseResult = tokenParser.CreateNode();
+
+            scope.AddNode(parseResult.node);
+            index = parseResult.index;
+
```
Comment (teusvanarkel) [inline NewParserTest/Parser.cs:32]: remove emtpy lines

File: NewParserTest/Parsers/OperationParser.cs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        _tokens = tokens;
+        _currentToken = currentToken;
+    }
+
+    public (INode node, int index) CreateNode()
+    {
+        INode node = null;
+
+        var index = _tokens.IndexOf(_currentToken);
+        INode LeftNode = new ValueNode(_tokens[index]);
```
Comment (teusvanarkel) [inline NewParserTest/Parsers/OperationParser.cs:25]: use var

File: NewParserTest/Parsers/OperationParser.cs [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    private List<Token> _tokens;
+    private Token _currentToken;
+
+    public OperationParser(List<Token> tokens, Token currentToken)
+    {
+        _tokens = tokens;
+        _currentToken = currentToken;
+    }
+
+    public (INode node, int index) CreateNode()
```
Comment (teusvanarkel) [inline NewParserTest/Parsers/OperationParser.cs:20]: Refactor

File: NewParserTest/Parsers/OperationParser.cs [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+                    index = result.index;
+
+                    node = new ComparisonOperationNode(LeftNode, OperatorNode, RightNode);
+                }
+            }
+        }
+        return (node, index);
+    }
+
+    public (INode node, int index) CreateBinaryOperationNode(int index, INode LeftNode)
```
Comment (teusvanarkel) [inline NewParserTest/Parsers/OperationParser.cs:85]: Think about better naming

File: NewParserTest/Parsers/VariableAssignParser.cs [hunk 11]
```diff
@@ -0,0 +1,52 @@
+using NewPirateLexer.Tokens;
+using NewParserTest.Parsers.Interfaces;
+using NewParserTest.Node.Interfaces;
+using NewPirateLexer.Enums;
+using NewParserTest.Node;
+
+namespace NewParserTest.Parsers;
+
+public class VariableAssignParser : ITokenParser
+{
```
Comment (teusvanarkel) [inline NewParserTest/Parsers/VariableAssignParser.cs:10]: Add OperationParser as dependency in constructor

File: NewParserTest/Parsers/VariableAssignParser.cs [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        var IdentifierNode = result.node;
+        index = result.index;
+
+        var Operator = _tokens[index += 1];
+        if (!Operator.Matches(TokenSyntax.EQUALS))
+        {
+            return (null, 0);
+        }
+
+        parser = parserFactory.GetParser(_tokens[index +=1], _tokens);
```
Comment (teusvanarkel) [inline NewParserTest/Parsers/VariableAssignParser.cs:39]: The parser doesnt need to be requested from the factory...itś already there

File: NewParserTest/Scope.cs [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using NewParserTest.Node.Interfaces;
+
+namespace NewParserTest
+{
+    public class Scope
+    {
+        public List<INode> Nodes { get; set; }
```
Comment (teusvanarkel) [inline NewParserTest/Scope.cs:11]: can be a public get, private set

File: NewParserTest/Scope.cs [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+        public bool AddNode(INode node)
+        {
+            try
+            {
+                Nodes.Append(node);
+                return true;
+            }
+            catch (System.Exception)
+            {
+                return false;
```
Comment (teusvanarkel) [inline NewParserTest/Scope.cs:22]: do something with the exception

**Review by teusvanarkel — DISMISSED** [review]
Litle improvements

File: NewParserTest/Node/ComparisonOperationNode.cs (hunk 6 above)
Comment (joerivanarkel) [reply NewParserTest/Node/ComparisonOperationNode.cs:8]: Properties are referenced in the interface

File: NewParserTest/Parsers/OperationParser.cs (hunk 9 above)
Comment (joerivanarkel) [reply NewParserTest/Parsers/OperationParser.cs]: Split Comparison and Binary Operations

File: Interpreter/Values/String.cs [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+namespace Interpreter.Values;
+
+public class String : BaseValue, IValue
+{
+    public override object Value { get; set; }
+    public Logger Logger { get; set; }
+
+    public String(object value, Logger logger)
+    {
+        Value = value;
```
Comment (teusvanarkel) [inline Interpreter/Values/String.cs:15]: Missing validation

File: Interpreter/Values/Variable.cs [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    public Variable(string value, Logger logger)
+    {
+        ValueNode = (IValueNode) SymbolTable.Instance(logger).Get(value);
+        Value= ValueNode.Value.Value;
+        Logger = logger;
+    }
+
+    public override BaseValue OperatedBy(Token _operator, BaseValue other)
+    {
+        var TypeCode = Type.GetTypeCode(Value.GetType());
```
Comment (teusvanarkel) [inline Interpreter/Values/Variable.cs:25]: Remove

File: Lexer/Error.cs [hunk 17]
```diff
@@ -0,0 +1,21 @@
+
+namespace Lexer
```
Comment (teusvanarkel) [inline Lexer/Error.cs:2]: remove or use

File: Lexer/Lexer.cs [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+namespace Lexer;
+public class Lexer
+{
+    public string fileName { get; set; }
+    public static string text { get; set; }
+    public static char currentChar { get; set; }
+    public static int position { get; set; }
+    public Logger Logger { get; set; }
+
+    public Lexer(string FileName, string Text, Logger logger)
```
Comment (teusvanarkel) [inline Lexer/Lexer.cs:16]: Make Logger an ILogger ( for testing later on )

File: Lexer/Lexer.cs [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+    {
+        Logger = logger;
```
Comment (teusvanarkel) [inline Lexer/Lexer.cs:18]: Add validation...if null throw an error

File: Lexer/TokenRepository.cs [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+        {
+            var idString = string.Empty;
+
+            while (Lexer.currentChar != null && (Globals.LETTERS_DIGITS.Contains(Lexer.currentChar) || Lexer.currentChar == '_'))
+            {
+                idString += Lexer.currentChar;
+                Lexer.Advance();
+            }
+
+            var tokenType = new object();
```
Comment (teusvanarkel) [inline Lexer/TokenRepository.cs:54]: Move to the top of the method

File: Lexer/TokenRepository.cs [hunk 21]
```diff
@@ -0,0 +1,276 @@
...
+            var typeKeywords = new string[] {"var", "int", "float", "string", "char", "new"};
+            var controlKeywords = new string[] {"if", "else", "for", "to", "foreach", "in", "while", "func", "class"};
+
+            if (typeKeywords.Contains(idString))
```
Comment (teusvanarkel) [inline Lexer/TokenRepository.cs:58]: Make a typeKeyWords manager which handles this

File: Lexer/TokenRepository.cs [hunk 22]
```diff
@@ -0,0 +1,276 @@
...
+            {
+                var tokenGroup = TokenGroup.TYPEKEYWORD;
```
Comment (teusvanarkel) [inline Lexer/TokenRepository.cs:60]: Make a tokegroupkeyword manager which handles this

File: Lexer/Tokens/Token.cs [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+    public TokenGroup TokenGroup { get; set; }
+    public object TokenType { get; set; }
+    public object? Value { get; set; }
+
+    public Token(TokenGroup tokenGroup, object tokenType, Logger logger, object value = null)
+    {
+        TokenGroup = tokenGroup;
+        TokenType = tokenType;
+        Value = value;
+        logger.Log($"Creating Token \"{TokenGroup.ToString()} | {TokenType.ToString()} | {(Value != null ? Value.ToString() : "None")}\"", this.GetType().Name, Common.Enum.LogType.INFO);
```
Comment (teusvanarkel) [inline Lexer/Tokens/Token.cs:18]: Add validation

Comment (teusvanarkel) [inline Lexer/Tokens/Token.cs:18]: Gebruik de ToString method

File: Parser/Parser.cs [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+
+    private List<Token> _tokens;
+    private IParserFactory parserFactory = new ParserFactory();
+    private ITokenParser tokenParser;
+    public Logger Logger { get; set; }
+    public ObjectSerializer ObjectSerializer { get; set; }
+    public string FileName { get; set; }
+    public Parser(List<Token> tokens, Logger logger, ObjectSerializer objectSerializer, string fileName)
+    {
+        _tokens = tokens;
```
Comment (teusvanarkel) [inline Parser/Parser.cs:22]: Validation

File: Parser/Parser.cs [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+        Logger = logger;
+        ObjectSerializer = objectSerializer;
+        FileName = fileName;
+        logger.Log("Created Parser", this.GetType().Name, LogType.INFO);
+    }
+    public Scope StartParse()
+    {
+        var index = 0;
+        Scope scope = new Scope(Logger);
```
Comment (teusvanarkel) [inline Parser/Parser.cs:31]: use var

File: NewLexerTest/TokenRepository.cs (hunk 4 above)
Comment (joerivanarkel) [reply NewLexerTest/TokenRepository.cs]: #42

File: NewLexerTest/Lexer.cs (hunk 3 above)
Comment (joerivanarkel) [reply NewLexerTest/Lexer.cs]: #42

File: NewParserTest/Parsers/OperationParser.cs (hunk 10 above)
Comment (joerivanarkel) [reply NewParserTest/Parsers/OperationParser.cs]: #43

File: Lexer/Error.cs (hunk 17 above)
Comment (joerivanarkel) [reply Lexer/Error.cs:2]: #44

File: Lexer/Lexer.cs (hunk 19 above)
Comment (joerivanarkel) [reply Lexer/Lexer.cs:18]: #46

File: Lexer/TokenRepository.cs (hunk 21 above)
Comment (joerivanarkel) [reply Lexer/TokenRepository.cs:58]: #47

File: Lexer/TokenRepository.cs (hunk 22 above)
Comment (joerivanarkel) [reply Lexer/TokenRepository.cs:60]: #47

File: Lexer/Tokens/Token.cs (hunk 23 above)
Comment (joerivanarkel) [reply Lexer/Tokens/Token.cs:18]: #46

File: Parser/Parser.cs (hunk 24 above)
Comment (joerivanarkel) [reply Parser/Parser.cs:22]: #46

**Comment by joerivanarkel** [comment]
**Summary**:
- Completely remade Parser
    - Using a Factory pattern get a Parser per token type. This is using the LL parser pattern.
    - Creates a Scope, which consits of a List of Nodes.
- Completely remade Interpreter
    - Using a Factory pattern get a Interpreter for each node.
    - Returns a BaseValue foreach node.
- Created a Object serializer, which takes a Scope and serializes it to binary.

**Pull request merged**